### PR TITLE
add xgboost tutorial

### PR DESCRIPTION
**Pull request by lostmygithubaccount**
- add xgboost tutorial
- run readme
- add workflow

# PR into Azure/azureml-examples

## Checklist

I have:

- [x] read and followed the contributing guidelines
- [x] ran `python readme.py` after making changes (per contributing guidelines)
- [x] added required testing (per contributing guidelines)

## Changes

- add distributed CPU xgboost tutorial noteobok

fixes #

### CODE REVIEW

File: tutorials/using-xgboost/0.Untitled.ipynb [hunk 1]
```diff
@@ -0,0 +1,311 @@
+{
```
Comment (dkmiller) [inline tutorials/using-xgboost/0.Untitled.ipynb:1]: Meta-comment: "untitled" isn't a great file name.

File: tutorials/using-xgboost/environment.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+channels:
+  - conda-forge
+dependencies:
+  - python=3.8
+  - pip
+  - pip:
+    - https://default7955323798.blob.core.windows.net/publictestwheels/azureml_core-0.1.0.0-py3-none-any.whl # azureml-core
```
Comment (dkmiller) [inline tutorials/using-xgboost/environment.yml:7]: Should point to packages in PyPI, not private blobs.

File: tutorials/using-xgboost/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# XGBoost
+
+description: learn how to use [XGBoost](https://github.com/dmlc/xgboost) distributed on Azure
```
Comment (dkmiller) [inline tutorials/using-xgboost/README.md:3]: Nit: capitalize + punctuation.

File: tutorials/using-xgboost/src/run.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import pandas as pd
+import xgboost as xgb
+import dask.dataframe as dd
+
+from distributed import Client
+from dask_mpi import initialize
+from adlfs import AzureBlobFileSystem
+
+# argparse setup
+parser = argparse.ArgumentParser()
```
Comment (dkmiller) [inline tutorials/using-xgboost/src/run.py:14]: This logic should go under an `if __name__ == "__main__"` block.

**Review by dkmiller — CHANGES_REQUESTED** [review]

File: tutorials/using-xgboost/src/run.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
+# imports
+import mlflow
+import argparse
+
+import pandas as pd
+import xgboost as xgb
+import dask.dataframe as dd
+
+from distributed import Client
+from dask_mpi import initialize
```
Comment (aminsaied) [inline tutorials/using-xgboost/src/run.py:10]: Nit: Prefer `import dask_mpi` and then use namespaces `dask_mpi.initialize()` - easier to read.

File: tutorials/using-xgboost/1.distributed-cpu.ipynb [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from azureml.core import ScriptRunConfig, Experiment, Environment\n",
+    "from azureml.core.runconfig import MpiConfiguration\n",
+    "\n",
+    "arguments = [\"--num_boost_round\", 100, \"--learning_rate\", 0.2, \"--gamma\", 0]\n",
```
Comment (pjanowski) [inline tutorials/using-xgboost/1.distributed-cpu.ipynb:55]: Split out things the user can set (like this but also especially the compute_target name) into a separate cell so user has clean vision of everything they can/need to set before running the training.

File: tutorials/using-xgboost/environment.yml (hunk 2 above)
Comment (lostmygithubaccount) [reply tutorials/using-xgboost/environment.yml:7]: agree, waiting on public release, will not merge until changed

File: tutorials/using-xgboost/0.Untitled.ipynb (hunk 1 above)
Comment (lostmygithubaccount) [reply tutorials/using-xgboost/0.Untitled.ipynb:1]: how is ``1.local-eda.ipynb``?

File: tutorials/using-xgboost/src/run.py (hunk 5 above)
Comment (lostmygithubaccount) [reply tutorials/using-xgboost/src/run.py:10]: went back and forth on that, it's not how horovod is shown in our examples is what I thought I saw

File: tutorials/using-xgboost/environment.yml (hunk 2 above)
Comment (lostmygithubaccount) [reply tutorials/using-xgboost/environment.yml:7]: [internal teams thread](https://teams.microsoft.com/l/message/19:c9f97cad7a7746789df07ebf82436fc3@thread.skype/1610403060948?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=88aa174e-6310-4634-bfcb-5761e1a1190a&parentMessageId=1610403060948&teamName=ML%20Platform&channelName=ASK%20MODE&createdTime=1610403060948)

File: tutorials/using-xgboost/src/run.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+parser.add_argument("--num_boost_round", type=int, default=10)
+parser.add_argument("--learning_rate", type=float, default=0.1)
+parser.add_argument("--gamma", type=float, default=0)
+parser.add_argument("--max_depth", type=int, default=8)
+args = parser.parse_args()
+
+# distributed setup
+print("initializing...")
+initialize()
+c = Client()
```
[the remark this answers is not in the document]
Comment (lostmygithubaccount) [reply tutorials/using-xgboost/src/run.py]: renamed to "client"

File: tutorials/using-xgboost/src/run.py (hunk 4 above)
Comment (lostmygithubaccount) [reply tutorials/using-xgboost/src/run.py:13]: given the discussion, should we do this for this example?

Comment (dkmiller) [reply tutorials/using-xgboost/src/run.py:13]: For a file this large (70+ lines) I stand by the comment (aka: yes).

File: tutorials/using-xgboost/src/run.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+import dask_mpi
+
+import xgboost as xgb
+import dask.dataframe as dd
+
+from distributed import Client
+from adlfs import AzureBlobFileSystem
+
+# begin script
+if __name__ == "__main__":
```
Comment (dkmiller) [inline tutorials/using-xgboost/src/run.py:13]: Suggested pattern is something like this:

```python
import argparse


def main(args):
    # interesting logic here


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("...")

    args = parser.parse_args()
    main(args)
````

This way others can e.g. import `main` from a Jupyter notebook or in unit tests.

**Review by dkmiller — CHANGES_REQUESTED** [review]

File: tutorials/using-xgboost/src/run.py (hunk 8 above)
Comment (lostmygithubaccount) [reply tutorials/using-xgboost/src/run.py]: updated, thanks!

**Review by dkmiller — DISMISSED** [review]

File: tutorials/using-xgboost/2.distributed-cpu.ipynb [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    "Simply use ``MpiConfiguration`` with the desired node count."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "nodes = 8 + 2\n",
```
Comment (mx-iao) [inline tutorials/using-xgboost/2.distributed-cpu.ipynb:52]: can you add a comment explaining this math

File: tutorials/using-xgboost/2.distributed-cpu.ipynb [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    "print(f\"Nodes: {nodes}\\nCPUs/node: {cpus_per_node}\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "arguments = {\n",
```
Comment (mx-iao) [inline tutorials/using-xgboost/2.distributed-cpu.ipynb:64]: why are you using a dict here?

Comment (lostmygithubaccount) [reply tutorials/using-xgboost/2.distributed-cpu.ipynb:64]: more logical to work with arguments in a dict than in a flat list

File: tutorials/using-xgboost/1.local-eda.ipynb [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+    "!pip install --upgrade dask==2.30.0 distributed==2.30.0 fastparquet adlfs xgboost pandas"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Get Data\n",
+    "\n",
+    "The data is modified from a kaggle competition and hosted publicly."
```
[the remark this answers is not in the document]
Comment (lostmygithubaccount) [reply tutorials/using-xgboost/1.local-eda.ipynb]: updated

File: tutorials/using-xgboost/2.distributed-cpu.ipynb (hunk 9 above)
Comment (lostmygithubaccount) [reply tutorials/using-xgboost/2.distributed-cpu.ipynb]: added markdown cell above + comments

**Review by mx-iao — APPROVED** [review]

File: tutorials/using-xgboost/2.distributed-cpu.ipynb (hunk 10 above)
Comment (lostmygithubaccount) [reply tutorials/using-xgboost/2.distributed-cpu.ipynb]: switched to list

**Review by dkmiller — APPROVED** [review]

**Pull request merged**